4.4-rc4: Windows resource compiler failed while compiling resources.rc.

4.4-rc4: Windows resource compiler failed while compiling a rc that contains #include "winres.h"

### Expected Behavior

### Current Behavior
```
C:\Users\fandre\Documents\git\gradle\subprojects\docs\src\samples\native-binaries\windows-resources>C:\ProgramFiles\gradle-4.4-rc-4\bin\gradle clean build
Support for running Gradle using Java 7 has been deprecated and is scheduled to be removed in Gradle 5.0. Please see https://docs.gradle.org/4.4-rc-4/userguide/java_plugin.html#sec:java_cross_compilation for more details.
Parallel execution is an incubating feature.
Support for nested build without a settings file was deprecated and will be removed in Gradle 5.0. You should create a empty settings file in C:\Users\fandre\Documents\git\gradle\subprojects\docs\src\samples\native-binaries\windows-resource
> Task :compileHelloSharedLibraryHelloRc FAILED
C:\Users\fandre\Documents\git\gradle\subprojects\docs\src\samples\native-binaries\windows-resources\src\hello\rc\resources.rc(2) : fatal error RC1015: cannot open include file 'winres.h'.


> Task :compileHelloStaticLibraryHelloRc FAILED
C:\Users\fandre\Documents\git\gradle\subprojects\docs\src\samples\native-binaries\windows-resources\src\hello\rc\resources.rc(2) : fatal error RC1015: cannot open include file 'winres.h'.



FAILURE: Build completed with 2 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':compileHelloSharedLibraryHelloRc'.
> A build operation failed.
      Windows resource compiler failed while compiling resources.rc.
  See the complete log at: file:///C:/Users/fandre/Documents/git/gradle/subprojects/docs/src/samples/native-binaries/windows-resources/build/tmp/compileHelloSharedLibraryHelloRc/output.txt

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

2: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':compileHelloStaticLibraryHelloRc'.
> A build operation failed.
      Windows resource compiler failed while compiling resources.rc.
  See the complete log at: file:///C:/Users/fandre/Documents/git/gradle/subprojects/docs/src/samples/native-binaries/windows-resources/build/tmp/compileHelloStaticLibraryHelloRc/output.txt

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

* Get more help at https://help.gradle.org

BUILD FAILED in 3s
7 actionable tasks: 7 executed

C:\Users\fandre\Documents\git\gradle\subprojects\docs\src\samples\native-binaries\windows-resources>
```

### Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

### Steps to Reproduce (for bugs)
Change the resource file by including: `#include "winres.h"`as below and build the project
```
C:\Users\fandre\Documents\git\gradle\subprojects\docs\src\samples\native-binaries\windows-resources>type src\hello\rc\resources.rc
#include "resources.h"
#include "winres.h"

STRINGTABLE
{
    IDS_HELLO,   "Hello world!"
}
```
### Your Environment
```
C:\Users\fandre\Documents\git\gradle\subprojects\docs\src\samples\native-binaries\windows-resources>C:\ProgramFiles\gradle-4.4-rc-4\bin\gradle -version

------------------------------------------------------------
Gradle 4.4-rc-4
------------------------------------------------------------

Build time:   2017-11-28 12:36:59 UTC
Revision:     f0a427cdb7eaa95779c9ccd8147878970fbb53fb

Groovy:       2.4.12
Ant:          Apache Ant(TM) version 1.9.9 compiled on February 2 2017
JVM:          1.7.0_80 (Oracle Corporation 24.80-b11)
OS:           Windows 8.1 6.3 amd64
```

--- a/subprojects/language-native/src/main/java/org/gradle/language/assembler/plugins/internal/AssembleTaskConfig.java
+++ b/subprojects/language-native/src/main/java/org/gradle/language/assembler/plugins/internal/AssembleTaskConfig.java
@@ -15,9 +15,13 @@
  */
 package org.gradle.language.assembler.plugins.internal;
 
+import com.google.common.collect.ImmutableSet;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.Project;
 import org.gradle.api.Task;
+import org.gradle.api.internal.file.FileCollectionFactory;
+import org.gradle.api.internal.file.collections.MinimalFileSet;
+import org.gradle.api.internal.project.ProjectInternal;
 import org.gradle.api.tasks.util.PatternSet;
 import org.gradle.internal.service.ServiceRegistry;
 import org.gradle.language.assembler.tasks.Assemble;
@@ -26,9 +30,15 @@ import org.gradle.language.base.internal.LanguageSourceSetInternal;
 import org.gradle.language.base.internal.SourceTransformTaskConfig;
 import org.gradle.nativeplatform.Tool;
 import org.gradle.nativeplatform.internal.NativeBinarySpecInternal;
+import org.gradle.nativeplatform.platform.internal.NativePlatformInternal;
+import org.gradle.nativeplatform.toolchain.internal.NativeToolChainInternal;
+import org.gradle.nativeplatform.toolchain.internal.PlatformToolProvider;
+import org.gradle.nativeplatform.toolchain.internal.SystemIncludesAwarePlatformToolProvider;
 import org.gradle.platform.base.BinarySpec;
 
 import java.io.File;
+import java.util.LinkedHashSet;
+import java.util.Set;
 
 public class AssembleTaskConfig implements SourceTransformTaskConfig {
     @Override
@@ -53,6 +63,23 @@ public class AssembleTaskConfig implements SourceTransformTaskConfig {
         task.setTargetPlatform(binary.getTargetPlatform());
 
         task.source(sourceSet.getSource());
+
+        FileCollectionFactory fileCollectionFactory = ((ProjectInternal) task.getProject()).getServices().get(FileCollectionFactory.class);
+        task.includes(fileCollectionFactory.create(new MinimalFileSet() {
+            @Override
+            public Set<File> getFiles() {
+                PlatformToolProvider platformToolProvider = ((NativeToolChainInternal) binary.getToolChain()).select((NativePlatformInternal) binary.getTargetPlatform());
+                if (platformToolProvider instanceof SystemIncludesAwarePlatformToolProvider) {
+                    return new LinkedHashSet<File>(((SystemIncludesAwarePlatformToolProvider) platformToolProvider).getSystemIncludes());
+                }
+                return ImmutableSet.of();
+            }
+
+            @Override
+            public String getDisplayName() {
+                return "System includes for " + binary.getToolChain().getDisplayName();
+            }
+        }));
 
         final Project project = task.getProject();
         task.setObjectFileDir(new File(binary.getNamingScheme().getOutputDirectory(project.getBuildDir(), "objs"), sourceSet.getProjectScopedName()));

--- a/subprojects/language-native/src/main/java/org/gradle/language/assembler/tasks/Assemble.java
+++ b/subprojects/language-native/src/main/java/org/gradle/language/assembler/tasks/Assemble.java
@@ -18,6 +18,7 @@ package org.gradle.language.assembler.tasks;
 import org.codehaus.groovy.runtime.DefaultGroovyMethods;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.Incubating;
+import org.gradle.api.file.ConfigurableFileCollection;
 import org.gradle.api.file.FileCollection;
 import org.gradle.api.tasks.Input;
 import org.gradle.api.tasks.InputFiles;
@@ -50,6 +51,7 @@ import java.util.concurrent.Callable;
 @Incubating
 public class Assemble extends DefaultTask {
     private FileCollection source;
+    private ConfigurableFileCollection includes;
     private NativeToolChainInternal toolChain;
     private NativePlatformInternal targetPlatform;
     private File objectFileDir;
@@ -58,6 +60,7 @@ public class Assemble extends DefaultTask {
     @Inject
     public Assemble() {
         source = getProject().files();
+        includes = getProject().files();
         getInputs().property("outputType", new Callable<String>() {
             @Override
             public String call() throws Exception {
@@ -83,6 +86,7 @@ public class Assemble extends DefaultTask {
 
         spec.setObjectFileDir(getObjectFileDir());
         spec.source(getSource());
+        spec.include(getIncludes());
         spec.args(getAssemblerArgs());
         spec.setOperationLogger(operationLogger);
 
@@ -150,5 +154,24 @@ public class Assemble extends DefaultTask {
 
     public void setObjectFileDir(File objectFileDir) {
         this.objectFileDir = objectFileDir;
+    }
+
+    /**
+     * Returns the header directories to be used for compilation.
+     *
+     * @since 4.4
+     */
+    @InputFiles
+    public FileCollection getIncludes() {
+        return includes;
+    }
+
+    /**
+     * Add directories where the compiler should search for header files.
+     *
+     * @since 4.4
+     */
+    public void includes(Object includeRoots) {
+        includes.from(includeRoots);
     }
 }

--- a/subprojects/language-native/src/main/java/org/gradle/language/rc/plugins/internal/WindowsResourcesCompileTaskConfig.java
+++ b/subprojects/language-native/src/main/java/org/gradle/language/rc/plugins/internal/WindowsResourcesCompileTaskConfig.java
@@ -15,10 +15,14 @@
  */
 package org.gradle.language.rc.plugins.internal;
 
+import com.google.common.collect.ImmutableSet;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.Project;
 import org.gradle.api.Task;
 import org.gradle.api.file.FileTree;
+import org.gradle.api.internal.file.FileCollectionFactory;
+import org.gradle.api.internal.file.collections.MinimalFileSet;
+import org.gradle.api.internal.project.ProjectInternal;
 import org.gradle.api.tasks.util.PatternSet;
 import org.gradle.internal.service.ServiceRegistry;
 import org.gradle.language.base.LanguageSourceSet;
@@ -29,9 +33,15 @@ import org.gradle.language.rc.tasks.WindowsResourceCompile;
 import org.gradle.nativeplatform.PreprocessingTool;
 import org.gradle.nativeplatform.internal.NativeBinarySpecInternal;
 import org.gradle.nativeplatform.internal.StaticLibraryBinarySpecInternal;
+import org.gradle.nativeplatform.platform.internal.NativePlatformInternal;
+import org.gradle.nativeplatform.toolchain.internal.NativeToolChainInternal;
+import org.gradle.nativeplatform.toolchain.internal.PlatformToolProvider;
+import org.gradle.nativeplatform.toolchain.internal.SystemIncludesAwarePlatformToolProvider;
 import org.gradle.platform.base.BinarySpec;
 
 import java.io.File;
+import java.util.LinkedHashSet;
+import java.util.Set;
 
 public class WindowsResourcesCompileTaskConfig implements SourceTransformTaskConfig {
     @Override
@@ -56,6 +66,24 @@ public class WindowsResourcesCompileTaskConfig implements SourceTransformTaskCon
         task.setTargetPlatform(binary.getTargetPlatform());
 
         task.includes(sourceSet.getExportedHeaders().getSourceDirectories());
+
+        FileCollectionFactory fileCollectionFactory = ((ProjectInternal) task.getProject()).getServices().get(FileCollectionFactory.class);
+        task.includes(fileCollectionFactory.create(new MinimalFileSet() {
+            @Override
+            public Set<File> getFiles() {
+                PlatformToolProvider platformToolProvider = ((NativeToolChainInternal) binary.getToolChain()).select((NativePlatformInternal) binary.getTargetPlatform());
+                if (platformToolProvider instanceof SystemIncludesAwarePlatformToolProvider) {
+                    return new LinkedHashSet<File>(((SystemIncludesAwarePlatformToolProvider) platformToolProvider).getSystemIncludes());
+                }
+                return ImmutableSet.of();
+            }
+
+            @Override
+            public String getDisplayName() {
+                return "System includes for " + binary.getToolChain().getDisplayName();
+            }
+        }));
+
         task.source(sourceSet.getSource());
 
         final Project project = task.getProject();

--- a/subprojects/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/app/MixedLanguageHelloWorldApp.groovy
+++ b/subprojects/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/app/MixedLanguageHelloWorldApp.groovy
@@ -110,10 +110,13 @@ class MixedLanguageHelloWorldApp extends HelloWorldApp {
         return i386GnuAsmSource
     }
 
+    // HACK: Ensure include root are correctly setup, use `INCLUDE <winbase.h>`
+    // See: https://github.com/gradle/gradle/issues/3662
     private static String windowsMasmSource = '''
 .386
 .model    flat
 
+INCLUDE <winbase.h>
 PUBLIC    _sumx
 _TEXT     SEGMENT
 _sumx    PROC

--- a/subprojects/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/app/MixedLanguageHelloWorldApp.groovy
+++ b/subprojects/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/app/MixedLanguageHelloWorldApp.groovy
@@ -110,13 +110,13 @@ class MixedLanguageHelloWorldApp extends HelloWorldApp {
         return i386GnuAsmSource
     }
 
-    // HACK: Ensure include root are correctly setup, use `INCLUDE <winbase.h>`
+    // HACK: Ensure include root are correctly setup, use `INCLUDE <ks386.inc>`
     // See: https://github.com/gradle/gradle/issues/3662
     private static String windowsMasmSource = '''
 .386
 .model    flat
 
-INCLUDE <winbase.h>
+INCLUDE <ks386.inc>
 PUBLIC    _sumx
 _TEXT     SEGMENT
 _sumx    PROC

--- a/subprojects/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/app/WindowsResourceHelloWorldApp.groovy
+++ b/subprojects/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/app/WindowsResourceHelloWorldApp.groovy
@@ -123,6 +123,10 @@ void DLL_FUNC hello() {
         sourceFile("rc", "resources.rc", """
 #include "hello.h"
 
+// HACK: Ensure include root are correctly setup
+// See: https://github.com/gradle/gradle/issues/3662
+#include "winres.h"
+
 STRINGTABLE
 {
     #ifdef FRENCH

